### ISSUE TITLE
Improvement: re-factor disabled worlds handling

### ISSUE THREAD

**Issue by jjchico** [sandain/MinecraftServerControlScript#76]
This is a proposal for improvement. I may do the implementation if I find the time. The idea is to discuss the details before implementation so that the final code can be integrated in the main repository.

## Rationale

Currently, enabled worlds are stored in `$LOCATION/worlds` and disabled worlds are moved to `$LOCATION/disabled-worlds`. So they are treated differently and it causes a number of glitches and bugs in the script derived from the fact that many functions of the script only consider enabled worlds. For example:

* Disabled worlds cannot be deleted: the have to be enabled first.
* Disabled worlds backups cannot be listed or restored.

In practice, a "disabled" world may be a world that is not started by default, but there is no reason to prevent any other operation on a disabled world, even manually starting the world's server.

## Proposal

Change the implementation of the disabled worlds feature in the following terms:

* `$LOCATION/worlds` would contain all the worlds data, both enabled and disabled.
* `$LOCATION/worlds-enabled`would contain symbolic links to enabled worlds. It is just a convenient way to store a list of enabled worlds.

### Required changes

A few changes in the code are needed but they would affect mostly world-handling methods in the script: `disableWorld()`, `getEnabledWorlds()` and the like. In most cases these methods are simplified by this change.

### Behaviour

* Any operation could be done on any world, enabled or not.
* Some operation on disabled worlds may required a "force" option. For example:
```
    $ mscs start karmaland
    mscs: 'karmaland' world is disabled. To force world start-up use:
    'mscs -f start karmaland'
```

**Comment by WebMaka** [comment]
Why continue using relocation and renaming at all? Since configuration files are already used on a per-world basis, it should be relatively simple to check each world's config for whether it should be enabled, e.g., append "mscs-enabled=true" to the properties file for enabled worlds or set it to false or even remove the line if disabled. That would put an end to all issues that pertain to relocation/renaming and should allow for streamlining for a few procedures since there would no longer be a need to play musical directories to find worlds. We could also use something like "mscs-clean-shutdown=false" (added during start, removed at the end of a proper shutdown) to catch dirty shutdowns instead of checking for $WORLDS_LOCATION/$1/$1-original.

**Comment by Shamus03** [comment]
I like both of your suggestions.  Details would need to be further discussed.

>append "mscs-enabled=true" to the properties file

This would be a good idea, but how long does it take to parse the properties file to find which worlds are enabled?  I can see this making bash completion sluggish, and it will definitely need to be rewritten to handle the change.  Of course, if the idea is to make every option available to all worlds, this becomes less important.  I'm just throwing out ideas.

As a whole, I think a line in the configuration is a cleaner and more responsible way of handling this.  If we want to remove the "musical directories" issue, we should do it entirely instead of bypassing the issue with symbolic links.

**Comment by jjchico** [comment]
The suggestion of using a configuration file is interesting and the methods to use one are already there. It does not affect the main idea of having a single folder to store all the worlds, enabled or not, and treat all the worlds the same, except for auto-start. How disabled worlds are identified is a second matter.

On the other hand, having mscs options mixed together with server options in server.properties may not be the best idea: what if a future version of the MC server decides to remove unknown options from server.properties? Having them in a separate mscs.properties file may be an option, not difficult too implement and will allow much faster processing of mscs options.

**Comment by zanix** [comment]
I suggested a separate mscs.properties file when we implemented the per-server properties.
One of the concerns was that minecraft would remove extra properties. I don't think a separate properties file would speed up scanning and parsing options.

**Comment by sandain** [comment]
I like the idea of moving all `mscs-` config options into their own file `mscs.properties`.  I'm not sure what my excuse was for keeping them in the `world.properties` file was anymore.

I also like the `mscs-enabled=true` option over the current situation, although I think that `true` should probably be the default value if the key or `mscs.properties` file is missing.

**Comment by zanix** [comment]
I don't recall the exact reason but I think it was "keep the file count low" or something

**Comment by jjchico** [comment]
It can be done in two steps:

1. Add mscs-enabled option to server.properties and change related code (quite a few places).
2. Move mscs options to a different file and change related code.

I already did some "improvements" on world listing that will really benefit from this. Since the new code is already working fine [1] I find it easier to make the changes on top of that branch, but I may work on master and integrate later if you prefer.

[1] https://github.com/jjchico/MinecraftServerControlScript/tree/better_world_lists

**Comment by sandain** [comment]
I went ahead and made this change, and I will upload the new code to my `properties` branch once I have split the code into reasonable patches (probably in the 2 step order that jjchico commented on).  Since this is a fairly drastic change, I won't commit it to the master until other people have had the chance to test it.  Sorry jjchico, I didn't base my changes on your patches, but it should be easy to merge your changes in once these go live.

**Comment by jjchico** [comment]
Great! I'll test it when I have a minute, and I will have a nice merge session with git :)

**Comment by sandain** [comment]
The original patches that I sent to the [properties](https://github.com/sandain/MinecraftServerControlScript/tree/properties) branch had a bug and was missing documentation.  I have now submitted new patches to fix these problems.

**Comment by jjchico** [comment]
Just ported my changes to your original properties branch, but it seems that you rebased/reset the original branch so it is quite a mess now. My changes are in [detailed_lists](https://github.com/jjchico/MinecraftServerControlScript/tree/detailed_lists) in case you want to check.

**Comment by sandain** [comment]
Sorry for the confusion jjchico.  I can cherry pick your changes tomorrow or sometime this weekend when I have the time.

**Comment by jjchico** [comment]
It is OK. I improved my Git skills and cherry-picked the changes myself. You can find it in the [lists](https://github.com/jjchico/MinecraftServerControlScript/tree/lists) branch (I will issue a PR in a moment).

With the new config format, most actions can be done on any available world, no matter if it is enabled or not: backup, restore, map, etc. It is already done for delete and status. With the getAvailableWorlds() method I implemented for lists the changes are trivial. I already updated mscs.completions to reflect that in my [dev](https://github.com/jjchico/MinecraftServerControlScript/tree/dev) branch but I better delay this PR until the lists changes are sorted out.

**Comment by jjchico** [comment]
I also added automatic world configuration migration for both disabled and enabled worlds, so user that upgrade the script will not have to do any manual editing. Every time the script is executed the config style is checked and migrated if necessary, transparently to the user.

I have done some checking but more testing would be needed.

I did the changes on top of my lists branch because I used some of the helper functions defined there, but the new commit is in a new branch [auto_config_migration](https://github.com/jjchico/MinecraftServerControlScript/tree/auto_config_migration).
